provider/openstack/local_test.go

```diff
@@ -2548,6 +2554,131 @@ func (s *localServerSuite) TestUpdateGroupController(c *gc.C) {
+
+	cl := client.NewClient(s.cred, identity.AuthUserPass, nil)
+	err := cl.Authenticate()
+	c.Assert(err, jc.ErrorIsNil)
+	containerURL, err := cl.MakeServiceURL("object-store", "", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	attrs := coretesting.FakeConfig().Merge(coretesting.Attrs{
+		"name":               "sample-no-neutron",
+		"type":               "openstack",
+		"auth-mode":          "userpass",
+		"agent-version":      coretesting.FakeVersionNumber.String(),
+		"agent-metadata-url": containerURL + "/juju-dist-test/tools",
+		"image-metadata-url": containerURL + "/juju-dist-test",
+		"authorized-keys":    "fakekey",
+	})
+	s.PatchValue(&jujuversion.Current, coretesting.FakeVersionNumber)
+	// For testing, we create a storage instance to which is uploaded tools and image metadata.
+	env, err := bootstrap.Prepare(
+		envtesting.BootstrapContext(c),
+		jujuclient.NewMemStore(),
+		prepareParams(attrs, s.cred),
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	s.env = env
+	s.toolsMetadataStorage = openstack.MetadataStorage(s.env)
+	// Put some fake metadata in place so that tests that are simply
+	// starting instances without any need to check if those instances
+	// are running can find the metadata.
+	envtesting.UploadFakeTools(c, s.toolsMetadataStorage, s.env.Config().AgentStream(), s.env.Config().AgentStream())
+	s.imageMetadataStorage = openstack.ImageMetadataStorage(s.env)
+	openstack.UseTestImageData(s.imageMetadataStorage, s.cred)
+	s.storageAdapter = makeMockAdapter()
+	overrideCinderProvider(c, &s.CleanupSuite, s.storageAdapter)
+}
+
+func (s *noNeutronSuite) TearDownTest(c *gc.C) {
+	if s.imageMetadataStorage != nil {
+		openstack.RemoveTestImageData(s.imageMetadataStorage)
+	}
+	if s.toolsMetadataStorage != nil {
+		envtesting.RemoveFakeToolsMetadata(c, s.toolsMetadataStorage)
+	}
+	s.srv.stop()
+	s.BaseSuite.TearDownTest(c)
+}
+
+func (s *noNeutronSuite) TestUpdateGroupControllerNoNeutron(c *gc.C) {
+	client := openstack.GetNovaClient(s.env)
+	_, err := client.CreateSecurityGroup("unrelated", "you know, for tests!")
+	c.Assert(err, jc.ErrorIsNil)
+	err = bootstrapEnv(c, s.env)
+	c.Assert(err, jc.ErrorIsNil)
+
+	groupNamesBefore := set.NewStrings(getNovaSecurityGroupNames(c, client)...)
+	c.Assert(groupNamesBefore, gc.DeepEquals, set.NewStrings(
+		"default",
+		"unrelated",
+		"juju-deadbeef-1bad-500d-9000-4b1d0d06f00d-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+		"juju-deadbeef-1bad-500d-9000-4b1d0d06f00d-deadbeef-0bad-400d-8000-4b1d0d06f00d-0",
+	))
+
+	firewaller := openstack.GetFirewaller(s.env)
+	err = firewaller.UpdateGroupController("aabbccdd-eeee-ffff-0000-0123456789ab")
+	c.Assert(err, jc.ErrorIsNil)
+
+	groupNamesAfter := set.NewStrings(getNovaSecurityGroupNames(c, client)...)
+	c.Assert(groupNamesAfter, gc.DeepEquals, set.NewStrings(
+		"default",
+		"unrelated",
+		"juju-aabbccdd-eeee-ffff-0000-0123456789ab-deadbeef-0bad-400d-8000-4b1d0d06f00d",
+		"juju-aabbccdd-eeee-ffff-0000-0123456789ab-deadbeef-0bad-400d-8000-4b1d0d06f00d-0",
```

Comment:
so is the idea that without the regex we used to fail to UpgradeGroup, or that we were modifying the old groups somehow? (I'm guessing the former).

maybe a comment in the test to clarify the specific reasons for this test.
// ensure that when Juju updates the security groups when we don't have
// Neutron networking, that we don't get confused by security groups
// that are not part of this model.

I also wonder if it would be useful to mix in Juju-* security groups which *aren't* the model we are updating. To make sure we don't touch those either (eg, the rule isn't just juju-.*)

Reply:
The former, yes - I'll add the comment. Also, other `juju-` groups for a different model is a good call, I'll add those too.